README.rst
Django Crowdsourced Fields
==========================

A reusable Django app that allows to mark certain fields of your models as
masterdata. Users would still be able to enter their own values but the app
will map them to unique instances. Admin staff is able to review all user
generated entries and mark them as approved.

An example could be a vehicle site, where you would like to allow users to
enter make and model for their vehicle but you want to make sure that an
entry of "bmw" and "Bmw" results in "BMW".

The app also comes with a nice jQuery combobox for such fields, where user get
auto-suggestions while they type.

Installation
------------

You need to install the following prerequisites in order to use this app::

    pip install Django
    pip install South

If you want to install the latest stable release from PyPi::

    $ pip install django-crowdsourced-fields

If you feel adventurous and want to install the latest commit from GitHub::

    $ pip install -e git://github.com/bitmazk/django-crowdsourced-fields.git#egg=crowdsourced_fields

Add ``crowdsourced_fields`` to your ``INSTALLED_APPS``::

    INSTALLED_APPS = (
        ...,
        'crowdsourced_fields',
    )

Don't forget to migrate your database::

    ./manage.py migrate crowdsourced_fields

Add jQuery and jQuery UI to your base template or at least to the template that 
should display forms with crowdsourced fields. Also include the jQuery UI
styles and special styles provided by this app::

    <link rel="stylesheet" type="text/css" href="http://ajax.googleapis.com/ajax/libs/jqueryui/1/themes/flick/jquery-ui.css">
    {{ form.media.css }}

    <script src="//ajax.googleapis.com/ajax/libs/jquery/1.7.0/jquery.min.js"></script>
    <script src="//ajax.googleapis.com/ajax/libs/jqueryui/1.8.21/jquery-ui.min.js"></script>
    {{ form.media.js }}

You might want to include the jQuery and jQuery UI parts in your base template
and the ``{{ form.media }}`` parts only in the template that uses a form with
crowdsourced fields.

Usage
-----

Prepare your models
+++++++++++++++++++

First you need to modify the model that should have crowdsourced fields::

    from crowdsourced_fields.models import CrowdSourcedModelMixin

    class YourModel(CrowdsourcedModelMixin, models.Model):
        CROWDSOURCED_FIELDS = {
            'make': {'item_type': 'makes', }
            'model': {'item_type': 'models', }
        }

        make = models.CharField(...)
        model = models.CharField(...)

``CROWDSOURCED_FIELDS`` is a dictionary of dictionaries. The main keys are the
fields that should be crowdsourced. This must be ``CharFields``.

The inner dictionary supports the following keys as settings:

1. **item_type (mandatory)**: The name of the group under which the data of
   this field should be grouped. Let's assume you have two models and both have
   a field ``country`` which should have access to the same data. By giving
   the same ``item_type`` for the field on both models, they will use the same
   set of crowdsourced data.

For each field that you selected, the mixin will dynamically add a method
called ``fieldname_crowdsourced`` to the model. Therefore we will save both,
the value that the user actually entered (in it's original field) and a link to 
the unique and approved value that we maintain through this app.

For your staff users it is save to change the values of the
``CrowdsourcedItem`` objects. Since you should use those in your templates,
any typo fixes would be reflected on your site immediately without the need
of a datamigration.

Create a model form
+++++++++++++++++++

Next you would create a ``ModelForm`` for your model with crowdsourced fields::

    from django import forms
    from crowdsourced_fields.forms import CrowdsourcedFieldsFormMixin
    from your_app.models import YourModel

    class YourModelForm(CrowdsourcedFieldsFormMixin, forms.ModelForm):
        class Meta:
            model = YourModel

The ``CrowdsourcedFieldsFormMixin`` will do the magic for you and add replace
the original form field (a text input) with a combobox that has all existing
values ready for autosuggest.

Create your template
++++++++++++++++++++

As mentioned above, first make sure that jQuery, jQuery UI and the form's media
is available in your template.

After that you can initiate the comboboxes like so::

    $(document).ready(function(){
        $('#id_country').combobox({
            source: {{ form.country_crowdsourced_values|safe }}
        });
    });

In this case ``country`` would be the name of the form field.


Contribute
----------

If you want to contribute to this project, please perform the following steps::

    # Fork this repository
    # Clone your fork
    $ mkvirtualenv -p python2.7 django-crowdsourced-fields
    $ pip install -r requirements.txt
    $ ./online_docs/tests/runtests.sh
    # You should get no failing tests

    $ git co -b feature_branch master
    # Implement your feature and tests
    $ ./crowdsourced_fields/tests/runtests.sh
    # You should still get no failing tests
    # Describe your change in the CHANGELOG.txt
    $ git add . && git commit
    $ git push origin feature_branch
    # Send us a pull request for your feature branch

Whenever you run the tests a coverage output will be generated in
``tests/coverage/index.html``. When adding new features, please make sure that
you keep the coverage at 100%.

If you are making changes that need to be tested in a browser (i.e. to the
CSS or JS files), you might want to setup a Django project, follow the
installation insttructions above, then run ``python setup.py develop``. This
will just place an egg-link to your cloned fork in your project's virtualenv.

Roadmap
-------

Check the issue tracker on github for milestones and features to come.


CHANGELOG.txt
=== 0.5 (ongoing) ===

=== 0.4 ===

* IMPORTANT: Fixed bad bug. It was not possible to have two fields on the same
  model that use the same item type. I had to add a new field to the generic
  key model, which is a mandatory field. After updating to this version,
  you need to delete all your items and generic FKs and create a datamigration
  that iterates over all your objects and saves them, so that the items
  get re-created. Unfortunately you will have to approve all items again.

=== 0.3  ===

* Fixed MANIFEST.in - now static files are included

=== 0.2 ===

* Bugfixes

=== 0.1 ===

Initial commit


crowdsourced_fields/tests/urls.py
"""
This ``urls.py`` is only used when running the tests via ``runtests.py``.
As you know, every app must be hooked into yout main ``urls.py`` so that
you can actually reach the app's views (provided it has any views, of course).

"""
from django.conf.urls.defaults import include, patterns, url
from django.contrib import admin


admin.autodiscover()


urlpatterns = patterns(
    '',
    url(r'^admin/', include(admin.site.urls)),
    url(r'^', include('crowdsourced_fields.tests.test_app.urls')),
)


crowdsourced_fields/tests/factories.py
"""Factories for the models of the ``django-crowdsourced-fields`` app."""
import factory

from crowdsourced_fields.models import CrowdsourcedItem


class CrowdsourcedItemFactory(factory.Factory):
    """Factory for the ``CrowdsourcedItem`` model."""
    FACTORY_FOR = CrowdsourcedItem

    item_type = 'countries'
    value = 'Germany'


crowdsourced_fields/tests/admin_tests.py
"""Tests for admin classes of the ``django-crowdsourced-fields`` app."""
from django.test import TestCase

from crowdsourced_fields.admin import approve_items
from crowdsourced_fields.models import CrowdsourcedItem
from crowdsourced_fields.tests.models_tests import create_fixtures


class ApproveItemsTestCase(TestCase):
    """Tests for the ``approve_items`` admin action."""
    longMessage = True

    def test_action(self):
        create_fixtures(self)
        user_generated_items = CrowdsourcedItem.objects.filter(
            is_user_generated=True)
        self.assertEqual(user_generated_items.count(), 3, msg=(
            'We assume that our fixtures have two user generated items'))
        queryset = CrowdsourcedItem.objects.filter(pk__in=[
            user_generated_items[0].pk, user_generated_items[1].pk])

        approve_items(None, None, queryset)
        user_generated_items = CrowdsourcedItem.objects.filter(
            is_user_generated=True)
        self.assertEqual(user_generated_items.count(), 1, msg=(
            'Should set `is_user_generated` to `False` for all selected'
            ' items'))


crowdsourced_fields/admin.py
"""Admin classes for the ``django-crowdsourced-fields`` app."""
from django.contrib import admin

from crowdsourced_fields.models import (
    CrowdsourcedItem,
    CrowdsourcedItemGenericForeignKey,
)


def approve_items(modeladmin, request, queryset):
    queryset.update(is_user_generated=False)
approve_items.short_description = "Approve selected items"


class CrowdsourcedItemAdmin(admin.ModelAdmin):
    list_display = ('item_type', 'value', 'is_user_generated', )
    list_filter = ('item_type', 'is_user_generated', )
    search_fields = ['value', ]
    actions = [approve_items]


class CrowdsourcedItemGenericForeignKeyAdmin(admin.ModelAdmin):
    list_display = ('item_type', 'item', 'content_type', 'object_id',
                    'field_name')
    list_filter = ('item_type', 'item__is_user_generated', 'field_name')
    search_fields = ['item__value', ]
    actions = [approve_items]


admin.site.register(CrowdsourcedItem, CrowdsourcedItemAdmin)
admin.site.register(
    CrowdsourcedItemGenericForeignKey, CrowdsourcedItemGenericForeignKeyAdmin)


crowdsourced_fields/forms.py
"""Forms for the ``django-crowdsourced-fields`` app."""
import json
import os

from django.conf import settings

from crowdsourced_fields.models import CrowdsourcedItem


def add_crowdsourced_values(cls, field_name, settings):
    """
    Dynamically adds a ``get_crowdsourced_values`` method to the form's fields.

    :cls: The form instance that should get the new method. Should be a formi
      instance with a model that has a ``CROWDSOURCED_FIELDS`` attribute.
    :field_name: The name of the original field.
    :settings: The dictionary with additional settings needed for adding the
      new method.

    """
    def inner_function(self):
        """
        The actual code of the function that returns the crowdsourced values.

        """
        this = inner_function
        this_name = this.__name__
        original_name = this_name.replace('_crowdsourced_values', '')
        model = self.__class__.Meta.model
        field_settings = model.CROWDSOURCED_FIELDS[original_name]
        items = list(CrowdsourcedItem.objects.filter(
            item_type=field_settings['item_type']).order_by(
                'value', ).values_list('value', flat=True))
        return json.dumps(items)

    inner_function.__doc__ = (
        'Returns the crowdsourced values for the {0} field'.format(
            field_name))
    inner_function.__name__ = '{0}_crowdsourced_values'.format(field_name)
    setattr(cls.__class__, inner_function.__name__, inner_function)


class CrowdsourcedFieldsFormMixin(object):
    class Media:
        css = {
            'all': (os.path.join(
                settings.STATIC_URL, 'crowdsourced_fields/css/combobox.css'), )
        }
        js = (
            os.path.join(
                settings.STATIC_URL, 'crowdsourced_fields/js/combobox.js', ),
        )

    def __init__(self, *args, **kwargs):
        super(CrowdsourcedFieldsFormMixin, self).__init__(*args, **kwargs)
        model = self.Meta.model
        for field_name, settings in model.CROWDSOURCED_FIELDS.items():
            add_crowdsourced_values(self, field_name, settings)
            instance = kwargs.get('instance')
            if instance:
                self.initial[field_name] = getattr(
                    instance, '{0}_crowdsourced'.format(field_name))()


crowdsourced_fields/models.py
"""Models of the ``django-crowdsourced-fields`` app."""
from django.contrib.contenttypes import generic
from django.contrib.contenttypes.models import ContentType
from django.db import models
from django.utils.translation import ugettext_lazy as _


def add_crowdsourced_method(cls, field_name, settings):
    """
    Dynamically adds a ``fieldname_crowdsourced`` method to an object.

    :cls: The object that should get the new method. Should be a model instance
      and should have a ``CROWDSOURCED_FIELDS`` attribute.
    :field_name: The name of the original field.
    :settings: The dictionary with additional settings needed for adding the
      new method.

    """
    def inner_function(self):
        """
        The actual code of the function that returns the crowdsourced value.

        """
        this = inner_function
        this_name = this.__name__
        original_name = this_name.replace('_crowdsourced', '')
        field_settings = self.CROWDSOURCED_FIELDS[original_name]
        item_type = field_settings['item_type']
        content_type = ContentType.objects.get_for_model(self)
        try:
            fk = CrowdsourcedItemGenericForeignKey.objects.get(
                content_type=content_type, object_id=self.pk,
                field_name=original_name, item_type=item_type)
        except CrowdsourcedItemGenericForeignKey.DoesNotExist:
            value = getattr(self, original_name)
            fk = self.get_crowdsourced_item(value, field_name, item_type)
        return fk.item.value

    inner_function.__doc__ = (
        'Returns the crowdsourced value for the {0} field'.format(
            field_name))
    inner_function.__name__ = '{0}_crowdsourced'.format(field_name)
    setattr(cls.__class__, inner_function.__name__, inner_function)


class CrowdsourcedFieldsModelMixin(object):
    """
    Adds ``fieldname_crowdsourced`` methods to the class.

    In order to use this mixin, please add a ``CROWDSOURCED_FIELDS`` attribute
    to your model class. See README for further instructions.

    Instead of displaying the real field in your templates, you should always
    display the crowdsourced value instead.

    """
    def __init__(self, *args, **kwargs):
        super(CrowdsourcedFieldsModelMixin, self).__init__(*args, **kwargs)
        for field_name, field_settings in self.CROWDSOURCED_FIELDS.items():
            add_crowdsourced_method(self, field_name, field_settings)

    def get_crowdsourced_item(self, original_value, field_name, item_type):
        """
        Returns the generic FK for a given original value.

        If no crowdsourced item and generic FK exists for the given original
        value, they will be created.

        :original_value: The value that the user has entered.
        :item_type: A string representing the item type of the value.

        """
        content_type = ContentType.objects.get_for_model(self)
        try:
            item = CrowdsourcedItem.objects.get(
                value__iexact=original_value, item_type=item_type)
        except CrowdsourcedItem.DoesNotExist:
            item = CrowdsourcedItem.objects.create(
                item_type=item_type, value=original_value)

        kwargs = {
            'content_type': content_type,
            'object_id': self.pk,
            'field_name': field_name,
            'item_type': item_type,
        }
        try:
            fk = CrowdsourcedItemGenericForeignKey.objects.get(**kwargs)
        except CrowdsourcedItemGenericForeignKey.DoesNotExist:
            fk = CrowdsourcedItemGenericForeignKey(**kwargs)

        if fk.item_id != item.pk:
            fk.item = item
            fk.save()
        return fk

    def save(self, *args, **kwargs):
        super(CrowdsourcedFieldsModelMixin, self).save(
            *args, **kwargs)
        for field_name, field_settings in self.CROWDSOURCED_FIELDS.items():
            item_type = field_settings['item_type']
            value = getattr(self, field_name).strip()
            self.get_crowdsourced_item(value, field_name, item_type)


class CrowdsourcedItem(models.Model):
    """
    Something that is both, masterdata and user entered data.

    For example a `country` field: Maybe you don't know which countries exist
    in the world so you want to allow users to enter anything. But you want
    to approve all entries and make sure that the spelling and capitalization
    is correct.

    :item_type: You might have a ``country`` field on several of your models
      and you want the same autosuggestions for all models. Therefore you need
      to define an ``item_type`` in your ``CROWDSOURCED_FIELDS`` attribute. By
      giving the same ``item_type`` to your models it means that the collected
      data is available to all these models.

    :value: The value that the user has entered.

    :is_user_generated: If ``True`` it means that this item has been created
      based on user input (and not by an admin). Admins should regularly check
      all user generated items for spelling errors and set this to ``False``
      once corrected and approved.

    """
    item_type = models.CharField(
        max_length=256,
        verbose_name=_('Item group'),
    )

    value = models.CharField(
        max_length=4000,
        verbose_name=_('Value'),
    )

    is_user_generated = models.BooleanField(
        default=True,
        verbose_name=_('Is user generated'),
    )

    def __unicode__(self):
        return '{0}: {1}'.format(self.item_type, self.value)


class CrowdsourcedItemGenericForeignKey(models.Model):
    """
    Maps ``CrowdsourcedItem`` objects to the objects that use them.

    This is a workaround because we don't want to dynamically add foreign
    keys to the objects that have crowdsourced fields. We use this mapping
    relationship instead to remember which object/field belongs to which
    ``CrowdsourcedItem``.

    :content_type: Part of the generic foreign key to the object this object
      belongs to.
    :object_id: See ``content_type``
    :content_object: See ``content_type``
    :field_name: A string representing the field name on the content object.
      This is needed if two fields use the same item type (i.e. ``Country of
      Birth`` and ``Nationality`` would both use the item type ``Countries``)
    :item: The crowdsourced item which holds the approved data.
    :item_type: We need this in order to know which ``CrowdsourcedItem`` to
      get.

    """
    # TODO: we have to wait for the unique constraint until next release,
    # see comment below

    content_type = models.ForeignKey(ContentType)
    object_id = models.PositiveIntegerField()
    content_object = generic.GenericForeignKey('content_type', 'object_id')

    # TODO: This should be required. It has blank=True so that we can get a
    # migration but in a later version we must require that users make sure
    # this field is always set before running the migration that finally sets
    # this to `blank=False`
    field_name = models.CharField(
        blank=True,
        max_length=256,
        verbose_name=_('Field name'),
    )

    item = models.ForeignKey(
        'crowdsourced_fields.CrowdsourcedItem',
    )

    item_type = models.CharField(
        max_length=256,
        verbose_name=_('Item group'),
    )


crowdsourced_fields/migrations/0002_auto__add_field_crowdsourceditemgenericforeignkey_field_name__del_uniq.py
# flake8: noqa
# -*- coding: utf-8 -*-
import datetime
from south.db import db
from south.v2 import SchemaMigration
from django.db import models


class Migration(SchemaMigration):

    def forwards(self, orm):
        # Removing unique constraint on 'CrowdsourcedItemGenericForeignKey', fields ['item_type', 'object_id']
        db.delete_unique('crowdsourced_fields_crowdsourceditemgenericforeignkey', ['item_type', 'object_id'])

        # Adding field 'CrowdsourcedItemGenericForeignKey.field_name'
        db.add_column('crowdsourced_fields_crowdsourceditemgenericforeignkey', 'field_name',
                      self.gf('django.db.models.fields.CharField')(default='', max_length=256, blank=True),
                      keep_default=False)


    def backwards(self, orm):
        # Deleting field 'CrowdsourcedItemGenericForeignKey.field_name'
        db.delete_column('crowdsourced_fields_crowdsourceditemgenericforeignkey', 'field_name')

        # Adding unique constraint on 'CrowdsourcedItemGenericForeignKey', fields ['item_type', 'object_id']
        db.create_unique('crowdsourced_fields_crowdsourceditemgenericforeignkey', ['item_type', 'object_id'])


    models = {
        'contenttypes.contenttype': {
            'Meta': {'ordering': "('name',)", 'unique_together': "(('app_label', 'model'),)", 'object_name': 'ContentType', 'db_table': "'django_content_type'"},
            'app_label': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'model': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'name': ('django.db.models.fields.CharField', [], {'max_length': '100'})
        },
        'crowdsourced_fields.crowdsourceditem': {
            'Meta': {'object_name': 'CrowdsourcedItem'},
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'is_user_generated': ('django.db.models.fields.BooleanField', [], {'default': 'True'}),
            'item_type': ('django.db.models.fields.CharField', [], {'max_length': '256'}),
            'value': ('django.db.models.fields.CharField', [], {'max_length': '4000'})
        },
        'crowdsourced_fields.crowdsourceditemgenericforeignkey': {
            'Meta': {'object_name': 'CrowdsourcedItemGenericForeignKey'},
            'content_type': ('django.db.models.fields.related.ForeignKey', [], {'to': "orm['contenttypes.ContentType']"}),
            'field_name': ('django.db.models.fields.CharField', [], {'max_length': '256', 'blank': 'True'}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'item': ('django.db.models.fields.related.ForeignKey', [], {'to': "orm['crowdsourced_fields.CrowdsourcedItem']"}),
            'item_type': ('django.db.models.fields.CharField', [], {'max_length': '256'}),
            'object_id': ('django.db.models.fields.PositiveIntegerField', [], {})
        }
    }

    complete_apps = ['crowdsourced_fields']


crowdsourced_fields/migrations/0001_initial.py
# flake8: noqa
# -*- coding: utf-8 -*-
import datetime
from south.db import db
from south.v2 import SchemaMigration
from django.db import models


class Migration(SchemaMigration):

    def forwards(self, orm):
        # Adding model 'CrowdsourcedItemGenericForeignKey'
        db.create_table('crowdsourced_fields_crowdsourceditemgenericforeignkey', (
            ('id', self.gf('django.db.models.fields.AutoField')(primary_key=True)),
            ('content_type', self.gf('django.db.models.fields.related.ForeignKey')(to=orm['contenttypes.ContentType'])),
            ('object_id', self.gf('django.db.models.fields.PositiveIntegerField')()),
            ('item', self.gf('django.db.models.fields.related.ForeignKey')(to=orm['crowdsourced_fields.CrowdsourcedItem'])),
            ('item_type', self.gf('django.db.models.fields.CharField')(max_length=256)),
        ))
        db.send_create_signal('crowdsourced_fields', ['CrowdsourcedItemGenericForeignKey'])

        # Adding unique constraint on 'CrowdsourcedItemGenericForeignKey', fields ['object_id', 'item_type']
        db.create_unique('crowdsourced_fields_crowdsourceditemgenericforeignkey', ['object_id', 'item_type'])

        # Adding model 'CrowdsourcedItem'
        db.create_table('crowdsourced_fields_crowdsourceditem', (
            ('id', self.gf('django.db.models.fields.AutoField')(primary_key=True)),
            ('item_type', self.gf('django.db.models.fields.CharField')(max_length=256)),
            ('value', self.gf('django.db.models.fields.CharField')(max_length=4000)),
            ('is_user_generated', self.gf('django.db.models.fields.BooleanField')(default=True)),
        ))
        db.send_create_signal('crowdsourced_fields', ['CrowdsourcedItem'])


    def backwards(self, orm):
        # Removing unique constraint on 'CrowdsourcedItemGenericForeignKey', fields ['object_id', 'item_type']
        db.delete_unique('crowdsourced_fields_crowdsourceditemgenericforeignkey', ['object_id', 'item_type'])

        # Deleting model 'CrowdsourcedItemGenericForeignKey'
        db.delete_table('crowdsourced_fields_crowdsourceditemgenericforeignkey')

        # Deleting model 'CrowdsourcedItem'
        db.delete_table('crowdsourced_fields_crowdsourceditem')


    models = {
        'contenttypes.contenttype': {
            'Meta': {'ordering': "('name',)", 'unique_together': "(('app_label', 'model'),)", 'object_name': 'ContentType', 'db_table': "'django_content_type'"},
            'app_label': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'model': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'name': ('django.db.models.fields.CharField', [], {'max_length': '100'})
        },
        'crowdsourced_fields.crowdsourceditem': {
            'Meta': {'object_name': 'CrowdsourcedItem'},
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'is_user_generated': ('django.db.models.fields.BooleanField', [], {'default': 'True'}),
            'item_type': ('django.db.models.fields.CharField', [], {'max_length': '256'}),
            'value': ('django.db.models.fields.CharField', [], {'max_length': '4000'})
        },
        'crowdsourced_fields.crowdsourceditemgenericforeignkey': {
            'Meta': {'unique_together': "(('object_id', 'item_type'),)", 'object_name': 'CrowdsourcedItemGenericForeignKey'},
            'content_type': ('django.db.models.fields.related.ForeignKey', [], {'to': "orm['contenttypes.ContentType']"}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'item': ('django.db.models.fields.related.ForeignKey', [], {'to': "orm['crowdsourced_fields.CrowdsourcedItem']"}),
            'item_type': ('django.db.models.fields.CharField', [], {'max_length': '256'}),
            'object_id': ('django.db.models.fields.PositiveIntegerField', [], {})
        }
    }

    complete_apps = ['crowdsourced_fields']
